# Segmenting and Clustering Neighborhoods in Toronto :

## Create and prepare Toronto postal codes Dataframe

### Imports and Initialization

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [11]:
# Import dependencies
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

Now assign the link of the website through which we are going to scrape the data and assign it to variable named website_url, and create a parse tree for parsed pages that can be used to extract data from HTML, which is useful for web scraping.

In [12]:
# getting data from aspecific source (wikipedia) 
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

#Find the required table are
soup = BeautifulSoup(website_url,'lxml')

# extracting the raw table inside that webpage
table = soup.find('table', {'class': 'wikitable sortable'})

### Data processing 1 :  Extract the table to a base dataframe
The base dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

In [13]:
table_pr = table.find_all('tr')

# Extract the table header (Extract three columns PostalCode, Borough, and Neighborhood)
table_header_pr = table_pr[0].findAll('th')
table_header = [str(th).replace('<th>', '').replace('</th>', '').replace('\n', '') for th in table_header_pr]

# Extract the table body
table_body_pr = table_pr[1:]
table_body = [list(map(lambda td: str(td if not td.findAll("a") 
                                         else td.find("a").getText()).replace('<td>', '')
                          .replace('</td>', '').replace('\n', '')
                          , tr.findAll('td'))) for tr in table_body_pr]


# Create a base dataframe to clean and transform
df_base = pd.DataFrame(table_body, columns=table_header)

### Data processing 2 :  Ignore cells with a borough that is not assigned

In [14]:
# replace "Not assigned" to NaN : To use dropna function for dropping the cells that have an Not assigned borough
df_base.replace("Not assigned", np.nan, inplace=True)

# Drop cells with "Not assigned" borough 
df_base.dropna(subset=["Borough"], inplace=True)

### Data processing 3 :  Replace Not assigned neighborhood by the borough
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [15]:
df_base.Neighbourhood.fillna(df_base.Borough, inplace=True)

### Data processing 4 :  Combine neighborhood that have the same postal code into one row
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [16]:
# Join neighboorhood Strings that have the same postal code to one row 
df_toronto = df_base.groupby(['Postcode', "Borough"])[["Neighbourhood"]].agg(lambda x: ', '.join(x)).reset_index()

## Visualize the appropriate pandas Dataframe

### 1. Visualizing the 10 first rows of the dataframe

In [17]:
df_toronto.head(10)

Postcode      Borough                                    Neighbourhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

### 2. Print the number of rows of your dataframe

In [18]:
df_toronto.shape

(103, 3)

##### The number of rows in the final dataframe equal to : 103

In [19]:
print("Number of rows in the dataframe : %d" % df_toronto["Postcode"].shape[0])

Number of rows in the dataframe : 103
